### ISSUE TITLE
Pages with `post-list` directives are not being rebuilt when adding new posts

### ISSUE THREAD

**Issue by chaghi** [getnikola/nikola#1671]
*Steps to reproduce*

* `nikola init test` (accept all the defaults)
* add a couple of posts
* add a page with a `post-list` directive
* `nikola build` (the site is built correctly for the first time)
* add more posts
* `nikola build`

*Expected result*

* Te page with the `post-list` directive should have been rebuilt to reflex the new posts.

*Actual result*

* Te page with the `post-list` is not rebuilt.

*Relevant thread on `nikola-discuss`*: https://groups.google.com/d/msg/nikola-discuss/ULhZ3mIG0yk/3740m_9ItLYJ

*Environment*

Ubuntu 14.04, Python 3.4.0, Nikola 7.3.1.1, with the following packages:
```console
blinker (1.3)
certifi (14.5.14)
docutils (0.12)
doit (0.27.0)
ghp-import (0.4.1)
ipython (3.1.0)
Jinja2 (2.7.3)
jsonschema (2.4.0)
livereload (2.3.2)
Logbook (0.9.0)
lxml (3.4.3)
Mako (1.0.1)
Markdown (2.6.2)
MarkupSafe (0.23)
micawber (0.3.2)
mistune (0.5.1)
natsort (3.5.6)
Nikola (7.3.1.1)
phpserialize (1.3)
Pillow (2.8.1)
pip (6.1.1)
ptyprocess (0.4)
pygal (1.7.0)
Pygments (2.0.2)
pyinotify (0.9.5)
Pyphen (0.9.1)
PyRSS2Gen (1.1)
python-dateutil (2.4.2)
pyzmq (14.6.0)
requests (2.6.2)
setuptools (15.1)
six (1.9.0)
smartypants (1.8.6)
terminado (0.5)
tornado (4.1)
typogrify (2.0.7)
Unidecode (0.4.17)
webassets (0.10.1)
Yapsy (1.11.23)
```

**Comment by ralsina** [comment]
First thing I found: you can't use post-list in a post because you get this:

````
ERROR: Cyclic/recursive dependencies for task render_posts:cache/posts/baz.html: [render_site -> render_posts -> render_posts:cache/posts/baz.html -> render_posts:cache/posts/baz.html]
````

Working to fix that then continuing :-)

**Comment by ralsina** [comment]
The core issue here is:

We are calculating the post list on rst-compile time. That will add dependencies on the latest posts *at that time*.

We add a new post -> nothing links it to the story using post-list. So it's not out of date, and nothing happens.

I will play a little to see if I find a reasonable solution.

**Comment by ralsina** [comment]
Second problem: if we try to use post.text() in post_list_directive.tmpl that fails horribly since the post may not be built yet. Working on that one now :-)

**Comment by ralsina** [comment]
Solved that one a bit brutishly...

**Comment by ralsina** [comment]
@chaghi you have a branch to test now, dude :-)

**Comment by chaghi** [comment]
@ralsina great! I've just did some testing on the branch. Seems ok! Pages with `post-list` directives are rebuilt as expected when new posts are created.

I've also played a bit with the "`page-list` directive inside a post" (not a page) scenario, and it also seems to work. I really wouldn't have expected the `post-list` directive to work inside posts (the circular dependency problem seemed logical), but hey! cool that now Nikola supports it! :-)

Thanks!
